snuba/subscriptions/consumer.py

```diff
@@ -122,6 +122,31 @@ def poll(self, timeout: Optional[float] = None) -> Optional[Message[Tick]]:
 
         return result
 
+    def tell(self) -> Mapping[Partition, int]:
+        # If there is no previous message for a partition, return the current
```

Comment:
With this we are establishing the convention that *the* offset of a Tick is the initial offset (since tell should return the next offset to be consumed, it should return the offset of the next Tick to be consumed, the Tick has two offsets so one of them has to be the offset that represents the tick). Is that correct?
I think it makes sense.

Reply:
Yes, I think so. It might be helpful to consider the invariants (from the caller's perspective) that I am trying to preserve here as well.

First, if a message is returned by `poll`, seeking to that message's offset should yield the _exact same message_, as demonstrated in the test by:

1. [`poll` returning a message in partition `1` with offset `1`](https://github.com/getsentry/snuba/blob/06aeb9e5e2f953bd7b95b17c653c191ea1350f7e/tests/subscriptions/test_consumer.py#L63-L68)
2. [`seek`ing in partition `1` to offset `1`](https://github.com/getsentry/snuba/blob/06aeb9e5e2f953bd7b95b17c653c191ea1350f7e/tests/subscriptions/test_consumer.py#L106) (Note that the [subsequent `poll` call returning `None`](https://github.com/getsentry/snuba/blob/06aeb9e5e2f953bd7b95b17c653c191ea1350f7e/tests/subscriptions/test_consumer.py#L119) shouldn't have any bearing on the consumer's behavior, there is no guarantee that every `poll` call will yield a message.)
3. [`poll` returning the same message payload as it did previously at that offset](https://github.com/getsentry/snuba/blob/06aeb9e5e2f953bd7b95b17c653c191ea1350f7e/tests/subscriptions/test_consumer.py#L132-L137)

Second, the working offset should never go backwards (an explicit `seek` call notwithstanding.) For example, this scenario should never occur:

```python
>>> consumer.tell()
{Partition(Topic("example"), 0): 10}
>>> message = consumer.poll()
>>> message.partition
Partition(Topic("example"), 0)
>>> message.offset
Partition(Topic("example"), 9)
```

Without the conditional here (on lines 133-135), the scenario described above would be possible if this was a direct proxy to `self.__consumer.tell()` since we need two messages to compose a `Tick`, and the offset of that `Tick` must be the lower of the two offsets (to maintain the first invariant.)

This is why the offsets by the outer (wrapping) consumer appear to be lagging by one message behind the offsets of the inner (wrapped) consumer in the tests, except immediately after assignment or seek.